usnic: Update the names of fabrics and domains

piroux opened the pull request:
Fix #2052 
Still WIP.

One question remains : what exactly should be done with the `strcmp` calls ?
Given that `fattrp->name` has been updated, it is no more equal to `dap->uda_devname`. However, this issue cannot be raised with the `fi_info` tool because this section of code is not reached in this tool.

```
if (fattrp->name != NULL &&
                    strcmp(fattrp->name, dap->uda_devname) != 0) { // CHANGE ??
            fprintf(stderr, "BOOM[1]\n");
            return -FI_ENODATA;
        }
```

The function `build_devname` will be cleaned once that issue has been resolved.

bturrubiates commented:
> However, this issue cannot be raised with the fi_info tool because this section of code is not reached in this tool.

Well if the compare is in place then it will break every call to `fi_fabric` meaning no tests will run successfully. I think we need to do some sort of comparison on the fabric name passed in. I'm not sure whether we should only match the exact name, or try to be a bit more friendly about it.

By that I mean: 

If we're exposing `10.5.0.0/16` as the fabric name and it corresponds to a device that has the IP address `10.5.20.195`, then should we allow fabric hints of the form `10.5.X.X/16` or only allow the string `10.5.0.0/16` as it is? We're going to have to generate the fabric name for each device as we iterate, or find some other comparison method if we aren't doing an exact `strcmp`. 

Thoughts? @jsquyres

I don't know if the filtering ever came up in the last discussion.

jsquyres commented:
@piroux Github pro tip: if you just have a naked #2052, it refers to issue number 2052 _in this repository_.  If you want to refer to an issue in a different repository, put its org/repo name in the front.  E.g., "Fixes ofiwg/libfabirc#2052" (the quotes in my example throw off the github UI parsing, but you get the idea).

jsquyres commented on prov/usnic/src/usdf_ep_dgram.c:
```diff
@@ -454,12 +454,16 @@ int usdf_dgram_fill_ep_attr(uint32_t version, struct fi_info *hints, struct
 	defaults = dgram_dflt_domain_attr;
-	defaults.name = strdup("usnic");
+
+	if (FI_VERSION_LE(version, FI_VERSION(1, 3)))
```
It would be worth putting a comment here explaining why we are doing this.  Future Pierre (and Future Ben and Future Jeff) will thank Past Pierre for reminding us why this version comparison is here, and what the two different behaviors are (and why).

piroux replied:
Yes it would be useful indeed.
Should I add references to issues ofiwg/libfabric#2052 and ofiwg/libfabric#1974 in this explanation ?

jsquyres commented on prov/usnic/src/usdf_ep_msg.c:
```diff
@@ -163,12 +163,16 @@ int usdf_msg_fill_ep_attr(struct fi_info *hints, struct fi_info *fi,
 	defaults = msg_dflt_domain_attr;
-	defaults.name = strdup("usnic");
+
+	if (FI_VERSION_LE(version, FI_VERSION(1, 3)))
```
Same comment for this version comparison / behavior difference.  If you want to put a lengthy comment / explanation in one place and have all the other places just refer to it, that would be fine (e.g., "See comment in foo.c explaining this version comparison and difference in behavior.").

bturrubiates replied:
Instead of defining `FI_VERSION_LE` you could always do: 

``` C
if (FI_VERSION_GE(version, FI_VERSION(1, 4)))
    ...
```

jsquyres commented:
@goodell and I had discussions about the "narrow match" and "wide match" of CIDR strings a while ago (in an OMPI context, but the issue is pretty much the same).  E.g., CIDR strings A and B clearly match when they each have the same contents.  But can you make the scope of the match "wider" than an exact/narrow match?  E.g., does "10.20.30.40/16" match "10.20.0.0/16", or "10.20.30.40/24"?

It quickly becomes a tangled web of sorrow.  My $0.02:
1. Keep it as a narrow match -- meaning that the strings match if `strcmp()` returns 0.
2. Keep the network-masked out bits of the IP address as 0 -- meaning that if the user submits the string "10.20.30.40/16", it would never match anything (because a /16 address would always be represented as "x.y.0.0" in libfabric).

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -117,7 +117,8 @@ usdf_validate_hints(struct fi_info *hints, struct usd_device_attrs *dap)
 		}
 		if (fattrp->name != NULL &&
-                    strcmp(fattrp->name, dap->uda_devname) != 0) {
+                    strcmp(fattrp->name, dap->uda_devname) != 0) { // CHANGE ??
```
I think you need to change this `strcmp()` to be based on the version, just like the others.  For <=v1.3, the current `strcmp()` is fine.  For >=v1.4, make up a CIDR string and compare against that.

Double check the sockets provider, but I assume they are doing something similar (at least for >=v1.4).

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -776,7 +824,8 @@ usdf_fabric_open(struct fi_fabric_attr *fattrp, struct fid_fabric **fabric,
 		dep = &dp->uu_info[d];
 		if (dep->ue_dev_ok &&
-			strcmp(fattrp->name, dep->ue_dattr.uda_devname) == 0) {
+			strcmp(fattrp->name, dep->ue_dattr.uda_devname) == 0) { // CHANGE ??
```
Same comment as with the other `strmp()` here.

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -189,6 +190,42 @@ usdf_fill_addr_info(struct fi_info *fi, uint32_t addr_format,
+	//USDF_DBG("NOW: ipaddr_be(network2):(%s)\n",
+	//	 addrnetw);
+
+	snprintf(devname, 40, "%s/%d", addrnetw, dap->uda_prefixlen);
```
Don't use a hard-coded 40 here -- use `sizeof(devname)`.

bturrubiates replied:
How about instead of using a local buffer + `strdup` you do something like: 

``` C
size_t size = snprintf(NULL, 0, "%s/%d", addrnetw, dap->uda_prefixlen) + 1;
*bufp = calloc(1, size);
sprintf(*bufp, "%s/%d", addrnetw, dap->uda_prefixlen);
```

jsquyres replied:
Meh.  I think a hard-coded 40 for the buffer length in the declaration/definition is fine -- you know that 40 bytes is definitely longer than you will need for this simple string, the max form of which is `xxx.xxx.xxx.xxx/xxx`.

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -189,6 +190,42 @@ usdf_fill_addr_info(struct fi_info *fi, uint32_t addr_format,
 	return ret;		// fi_freeinfo() in caller frees all
 }
 
+void build_devname(struct usd_device_attrs *dap, char **bufp)
```
My $0.02: if you have a function named `build_devname()`, that implies to me that it's going to build the devname for _all_ versions.  I.e., you shouldn't only call this function for >=v1.4.

Specifically: it might be a little better down below to _always_ call this function (and maybe name this function `get_getname()`?), and then have the version comparison / behavior difference in this function.

goodell replied:
> It quickly becomes a tangled web of sorrow. My $0.02:
> 1. Keep it as a narrow match -- meaning that the strings match if strcmp() returns 0.
> 2. Keep the network-masked out bits of the IP address as 0 -- meaning that if the user submits the string "10.20.30.40/16", it would never match anything (because a /16 address would always be represented as "x.y.0.0" in libfabric).

👍  on this approach.  Networks should only match if the mask is the same length (`/16` can only match `/16`, never `/22`).  Furthermore, apply the mask before making the address comparison otherwise we're just asking our clients to replicate that logic in lots of other software when it should be centralized in ours.

The "wide match" stuff was always nonsense that we only hypothetically discussed as a way to try to simplify the command line UI of Open MPI's `mpirun` program.  I don't think it ever made much sense.

piroux commented:
> > It quickly becomes a tangled web of sorrow. My $0.02:
> > 1. Keep it as a narrow match -- meaning that the strings match if strcmp() returns 0.
> > 2. Keep the network-masked out bits of the IP address as 0 -- meaning that if the user submits the string "10.20.30.40/16", it would never match anything (because a /16 address would always be represented as "x.y.0.0" in libfabric).
> 
> :+1: on this approach. Networks should only match if the mask is the same length (/16 can only match /16, never /22). Furthermore, apply the mask before making the address comparison otherwise we're just asking our clients to replicate that logic in lots of other software when it should be centralized in ours.
> 
> The "wide match" stuff was always nonsense that we only hypothetically discussed as a way to try to simplify the command line UI of Open MPI's mpirun program. I don't think it ever made much sense.

@jsquyres @goodell 
I am not sure to have understood the expected behaviour.

If a user submits an address like `10.20.30.40/16`, do you want to :
- reject it because the user should have submitted `10.20.0.0/16`.
- OR apply the netmask on it, so it becomes `10.20.0.0/16` and it will match ?

goodell commented:
> If a user submits an address like 10.20.30.40/16, do you want to :
> - reject it because the user should have submitted 10.20.0.0/16.
> - OR apply the netmask on it, so it becomes 10.20.0.0/16 and it will match ?

The second option (apply mask and match).

bturrubiates commented:
> The second option (apply mask and match).

👍  I think I like this option more.

jsquyres commented:
> The second option (apply mask and match).

👍

jithinjose committed:
common/proto: Add message types for RMA operations

- Minor updates for supporting RMA operations
  - Add read-req and read-resp
  - Add peer msg id

Signed-off-by: Jithin Jose <jithin.jose@intel.com>

jithinjose committed:
prov/udp: Fix memory leaks in UDP CQ

- CQ structure was not freed. This patch fixes the memory leak.

Signed-off-by: Jithin Jose <jithin.jose@intel.com>

jithinjose committed:
prov/rxd: Coverity scan fixes

- This commit fixes issues reported by Coverity scan and couple of
  memory leak issues in RxD provider.

Signed-off-by: Jithin Jose <jithin.jose@intel.com>

jithinjose committed:
prov/rxd: Add RMA read-write functionality

- This commit adds RMA read/write capabilities for RxD provider.

RMA write/write-data: Initiator sends 'ofi_write' header packet to
target process. The header packet includes the MR key, target IOVs and
data. Receiver verifies MR key, and the data is placed directly into
target memory. ACK packet for the last data packet indicates completion
of RMA transfer. RxD uses util-mr for keeping memory registrations, and
it follows the same flow control mechanisms that are used for send/recv.

RMA read: Initiator sends 'ofi_read_req' header packet, which includes
target MR key, and source IOVs. Receiver verifies the MR key, and sends
the data back to the initiator as 'ofi_read_rsp' packet. Initiator's
tx-entry ID is sent along with the ofi_read_rsp packet. Incoming data is
placed to IOVs that are stored in the initiator's tx-entry. The final
ACK for ofi_read_rsp indicates completion of RMA operation.

Signed-off-by: Jithin Jose <jithin.jose@intel.com>

piroux commented:
This PR currently includes commits for the `fi_pingpong` utility in order to test the behaviour of the name update.
However, the only relevant commit is the last one : `usnic: change of fabric/domain naming`.
Once the PR is ready for merging, the history will be cleaned.

Commit:
fixed windows build

fixed build on windows platform
- disabled processing of object in sock_wait_get_obj for
  windows platform
- updated Linux specific header replacer to provide compilation

Signed-off-by: Oblomov, Sergery <sergey.oblomov@intel.com>

jsquyres commented on prov/usnic/src/usdf.h:
```diff
@@ -47,7 +47,7 @@
 #define USDF_PROV_NAME "usnic"
 #define USDF_MAJOR_VERS 1
-#define USDF_MINOR_VERS 0
+#define USDF_MINOR_VERS 3
```
What is this version number change for?

piroux replied:
OK. I was confused with all the different kinds of versions available.

jsquyres commented on prov/usnic/src/usdf_domain.c:
```diff
@@ -344,3 +344,15 @@ usdf_domain_open(struct fid_fabric *fabric, struct fi_info *info,
+	if (FI_VERSION_GE(version, FI_VERSION(1, 4)))
+		bufp = strdup(dap->uda_devname);
+	else
+		bufp = strdup("usnic");
```
Per https://github.com/ofiwg/libfabric/issues/2052, shouldn't `bufp` return NULL when `<= v1.3`?  (i.e., revert the behavior of ofiwg/libfabric@ca81e01823ffcfe26e1a18f5f85eaf57f1fba591)

piroux replied:
OK

jsquyres commented on prov/usnic/src/debugme_func.c:
```diff
@@ -0,0 +1,100 @@
+
```
I assume that this whole file is just for debugging and will not be included in the final PR.

piroux replied:
Yes

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -75,12 +75,82 @@
 
 struct usdf_usnic_info *__usdf_devinfo;
 
+char *
```
Shouldn't this be `static`?

piroux replied:
OK

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -75,12 +75,82 @@
+		in.s_addr = dap->uda_ipaddr_be & dap->uda_netmask_be;
+		addrnetw = inet_ntoa(in);
+		size = snprintf(NULL, 0, "%s/%d", addrnetw, dap->uda_prefixlen) + 1;
+		bufp = calloc(1, size);
```
Need to check for `bufp` to ensure that `calloc()` didn't fail.

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -75,12 +75,82 @@
+		addrnetw = inet_ntoa(in);
+		size = snprintf(NULL, 0, "%s/%d", addrnetw, dap->uda_prefixlen) + 1;
+		bufp = calloc(1, size);
+		sprintf(bufp, "%s/%d", addrnetw, dap->uda_prefixlen);
```
Why not use `asprintf()` for this whole block?

``` c
asprintf(&bufp, "%s/%d", inet(ntoa(in)), dap->uda_prefixlen))
```

piroux replied:
Is it ok even if `asprintf` is not POSIX ?

The man page specifies :

```
CONFORMING TO
       These functions are GNU extensions, not in C or POSIX.  They are also available under *BSD.
```

jsquyres replied:
Yes.  We use `asprintf()` elsewhere in libfabric.

piroux replied:
OK

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -75,12 +75,82 @@
+	CHECK_DEVNAME_FABRIC
+};
+
+int
```
This should also be `static`.

piroux replied:
OK

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -75,12 +75,82 @@
+	char *devname;
+
+	char* devname_1_3 = get_devname(FI_VERSION(1,3), dap);
+	char* devname_1_4 = get_devname(FI_VERSION(1,4), dap);
```
You should only call these `get_devname()` for each version when/if you need to.  I.e., put them inside the version checks, below.

piroux replied:
OK

bturrubiates replied:
I talked it over with @piroux and decided that the functions should probably be named `check_fabric_name` and `get_fabric_name` since they aren't really tied to the device name at all.

Also we realized an insight here is that the version information will never be available in a call from `fi_fabric`, but will be available in a call from `fi_getinfo`. So if the version is available, we can just skip and check the version provided. Otherwise, we can check if there is a slash in the name and treat it appropriately. 

``` c
bool check_fabric_name(uint32_t version, struct usd_device_attrs *dap, char *hint)
{
    bool valid = false;
    char *reference;

    USDF_DBG("checking devname: version=%d, devname='%s'\n", version,
            hint);

    if (version) {
        reference = get_fabric_name(version, dap);
        if (strmp(reference, hint) == 0)
            valid = true;
        else
            USDF_DBG("hint %s failed to match %s\n", hint,
                    reference);

        free(reference);
        return valid;
    }

    /* The hint string itself is kind of a version check, in pre-1.4 the
     * name was just 'usnic'. In 1.4 and beyond, then name is actually CIDR
     * notation.
     */
    if (strstr(hint, '/'))
        return check_fabric_name(FI_VERSION(1, 4), dap, hint);

    return check_fabric_name(FI_VERSION(1, 3), dap, hint);
}
```

I find this to be clearer, thoughts @jsquyres?

jsquyres replied:
👍

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -75,12 +75,82 @@
+			break;
+	}
+
+
```
Remove the extra blank line here.

piroux replied:
OK

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -75,12 +75,82 @@
+	char *devname;
+
+	char* devname_1_3 = get_devname(FI_VERSION(1,3), dap);
+	char* devname_1_4 = get_devname(FI_VERSION(1,4), dap);
```
Meh; I see the overall logic of what you did here now -- i.e., it's ok to leave these `get_devname()`s up top.  But you should put a comment here explaining why this function exists (i.e., that we changed the device name between v1.3 and v1.4).

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -75,12 +75,82 @@
+	if (ret == 0)
+		return 1;
+	else
+		return 0;
```
Change this to:

``` c
return (ret == 0) ? 1 : 0;
```

piroux replied:
OK.
Indeed much better :+1:

jsquyres commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -116,13 +186,19 @@ usdf_validate_hints(struct fi_info *hints, struct usd_device_attrs *dap)
-                    strcmp(fattrp->name, dap->uda_devname) != 0) {
-			return -FI_ENODATA;
+                    !check_devname(version, dap, NULL, CHECK_DEVNAME_GETINFO)) {
+			USDF_DBG("devname of fabric not matching expected for version (version=%d) (devname='%s')\n", version, devn);
```
Change the message to:

```
for API version 0x%x, this device's name is %s; hint requested %s -- no match
```

piroux replied:
The `hint requested %s` part is a bit problematic because it is version dependent ... That is why I encapsulated the fetching and the checking of the device name in two functions.

So instated, what do you think of :
`for API version 0x%x, this device's name is %s; device requested in hints does not match`

bturrubiates replied:
Move this message into the `check_devname` function.  Log that they don't match using debug log level.

jsquyres replied:
👍

jsquyres commented:
Should probably also add a statement to `fi_usnic.7.md` about the change in fabric/domain names.

jsquyres commented:
Put "Fixes ofiwg/libfabric#2052" in the commit message so that it will auto-close the issue when the PR is merged.

I've filed two other PRs:
- https://github.com/ofiwg/libfabric/pull/2279: Update libfabric to v1.4
- https://github.com/open-mpi/ompi/pull/1989: Open MPI support for the new fabric/device names

So let's fix up the issues in this PR and file a final PR against the main libfabric project ASAP.  Thanks.

bturrubiates commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -75,12 +75,82 @@
+	if (ret == 0)
+		return 1;
+	else
+		return 0;
```
> Change this to:
> 
> return (ret == 0) ? 1 : 0;

Actually to: 

``` c
return (ret == 0); 
```

Commit:
prov/sockets: Return correct tx size when fi_tx_size_left() is called

Currently, the allocated tx size is rounded up to the power of two, fix to code to return actual
tx/rx size, added checking to handle the rounding up.

Signed-off-by: Shantonu Hossain <shantonu.hossain@intel.com>

Commit:
manpage/fi_av.3.md: Clarified fi_av_remove when garbage collection of addresses is implemented

If the garbage collection of av addresses is implemented by the provider,
addresses released by fi_av_remove can be referenced later if it is returned by
a new fi_av_insert call.

Signed-off-by: Shantonu Hossain <shantonu.hossain@intel.com>

Commit:
.gitignore: Add configure generated file

Signed-off-by: Ken Raffenetti <raffenet@mcs.anl.gov>

bturrubiates commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -116,13 +188,19 @@ usdf_validate_hints(struct fi_info *hints, struct usd_device_attrs *dap)
 			return -FI_ENODATA;
 		}
+
+		devn = get_devname(version, dap);
```
This isn't used here, you can get rid of it.

piroux replied:
OK

bturrubiates commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -116,13 +188,19 @@ usdf_validate_hints(struct fi_info *hints, struct usd_device_attrs *dap)
 		if (fattrp->name != NULL &&
-                    strcmp(fattrp->name, dap->uda_devname) != 0) {
-			return -FI_ENODATA;
+                    !check_devname(version, dap, NULL, CHECK_DEVNAME_GETINFO)) {
```
You should be passing the `fattrp` variable in the third argument. Maybe you should make that take a string hint instead?

piroux committed:
usnic: change of fabric/domain naming

Fixes ofiwg/libfabric#2052

Signed-off-by: Pierre Roux <piroux@cisco.com>

ofiwg-bot committed:
Updated nroff-generated man pages

j-xiong committed:
Merge pull request #2285 from shantonu/pr/manpage

manpage/fi_av.3.md: Clarified fi_av_remove

bturrubiates commented:
I've been thinking about the way that `usdf_domain_getname` was implemented and I think we actually have to implement it differently because there isn't a way to do effective error handling. So in the current design:

``` c
char *usdf_domain_getname(uint32_t version, struct usd_device_attrs *dap)
{
    char *bufp;
    if (FI_VERSION_GE(version, FI_VERSION(1, 4)))
        bufp = strdup(dap->uda_devname);
    else
        bufp = NULL;
    return bufp;
}
```

There isn't a way to determine whether the name is `NULL`, or if there was an error in the `strdup` call. Maybe we should change it to something like:

``` c

int usdf_domain_getname(uint32_t version, struct usd_device_attrs *dap,
        char **name)
{
    int ret = FI_SUCCESS;
    char *buf = NULL;

    if (FI_VERSION_GE(version, FI_VERSION(1, 4))) {
        buf = strdup(dap->uda_devname);
        if (!buf)
            ret = -errno;
    }

    *name = buf;
    return ret;
}
```

And of course make the appropriate changes to `get_fabric_name` to reflect this as well. Also we should probably consolidate the names, so maybe `usdf_domain_getname` and `usdf_fabric_getname`. 

@piroux

bturrubiates commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -75,9 +76,56 @@
+	}
+
+	/* The hint string itself is kind of a version check, in pre-1.4 the
+	* name was just 'usnic'. In 1.4 and beyond, then name is actually CIDR
```
Also, my fault this comment is wrong. In pre-1.4 the name wasn't just 'usnic' it was the device name.

bturrubiates commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -116,13 +164,14 @@ usdf_validate_hints(struct fi_info *hints, struct usd_device_attrs *dap)
 		}
 	}
 
-	return 0;
```
Make this `FI_SUCCESS`.

j-xiong committed:
Merge pull request #2250 from jithinjosepkl/pr/rxd

prov/rxd: Add RMA read-write functionality

j-xiong committed:
Merge pull request #2275 from shantonu/pr/sockets-size_left

Pr/sockets size left

j-xiong committed:
Merge pull request #2286 from raffenet/gitignore

.gitignore: Add configure generated file

bturrubiates commented:
@jsquyres I asked @piroux to update the PR style to be consistent with `checkpatch.pl` and libfabric core. I've requested that he also update any function prototypes/definitions that he modified.

j-xiong committed:
Merge pull request #2264 from hoopoepg/win_build_fix

fixed windows build

bturrubiates commented on prov/usnic/src/usdf_fabric.c:
```diff
@@ -75,8 +76,74 @@
+			       char **name)
+{
+	int ret = FI_SUCCESS;
+	char *bufp;
```
Initialize this to `NULL`.

bturrubiates commented on man/fi_usnic.7.md:
```diff
@@ -283,6 +283,24 @@ is the destination host is locally connected, and `1` otherwise.
+remains backward compatible with applications supporting the old scheme.
+
+Specifically, the usNIC provider now implements both conventions, and decides which one to use based on
+the version passed to the first call to "fi_getinfo":
```
put `fi_getinfo` in ` characters.

bturrubiates commented on man/fi_usnic.7.md:
```diff
@@ -283,6 +283,24 @@ is the destination host is locally connected, and `1` otherwise.
+Specifically, the usNIC provider now implements both conventions, and decides which one to use based on
+the version passed to the first call to "fi_getinfo":
+
+* When FI_VERSION(1,4) or higher is used:
```
put the `FI_VERSION(1,4)` in ` characters.

bturrubiates commented on man/fi_usnic.7.md:
```diff
@@ -283,6 +283,24 @@ is the destination host is locally connected, and `1` otherwise.
+     - fabric name is the network address with the CIDR notation (i.e., a.b.c.d/e)
+     - domain name is the usNIC Linux interface name (i.e., usnic_X)
+
+* When a lower version number is used, like FI_VERSION(1, 3), it follows the same behavior the usNIC provider exhibited in libfabric <= v1.3:
```
put the `FI_VERSION(1,3)` in ` characters.

bturrubiates commented on man/fi_usnic.7.md:
```diff
@@ -283,6 +283,24 @@ is the destination host is locally connected, and `1` otherwise.
+
+* When a lower version number is used, like FI_VERSION(1, 3), it follows the same behavior the usNIC provider exhibited in libfabric <= v1.3:
+     - fabric name is be the usNIC Linux interface name (i.e., usnic_X)
+     - domain name is NULL
```
`NULL` in `.

bturrubiates commented on man/fi_usnic.7.md:
```diff
@@ -283,6 +283,24 @@ is the destination host is locally connected, and `1` otherwise.
+the version passed to the first call to "fi_getinfo":
+
+* When FI_VERSION(1,4) or higher is used:
+     - fabric name is the network address with the CIDR notation (i.e., a.b.c.d/e)
```
put a.b.c.d/e in `.

bturrubiates commented on man/fi_usnic.7.md:
```diff
@@ -283,6 +283,24 @@ is the destination host is locally connected, and `1` otherwise.
+
+* When FI_VERSION(1,4) or higher is used:
+     - fabric name is the network address with the CIDR notation (i.e., a.b.c.d/e)
+     - domain name is the usNIC Linux interface name (i.e., usnic_X)
```
put the `usnic_X` in `.

bturrubiates commented on man/fi_usnic.7.md:
```diff
@@ -283,6 +283,24 @@ is the destination host is locally connected, and `1` otherwise.
+The release of libfabric v1.4 introduced a new naming convention for fabric and domain. However the usNIC provider
+remains backward compatible with applications supporting the old scheme.
+
+Specifically, the usNIC provider now implements both conventions, and decides which one to use based on
```
Remove this sentence

bturrubiates commented on man/fi_usnic.7.md:
```diff
@@ -283,6 +283,24 @@ is the destination host is locally connected, and `1` otherwise.
+## New naming convention for fabric/domain starting with libfabric v1.4
+
+The release of libfabric v1.4 introduced a new naming convention for fabric and domain. However the usNIC provider
+remains backward compatible with applications supporting the old scheme.
```
Add "and decides which one to use based on the version passed to `fi_getinfo`" here.

bturrubiates commented:
@jsquyres this looks good to me (after pending `fi_usnic.7.md` fixes), I think it's ready for upstream PR and merge. What do you think?

jsquyres commented:
👍
